zerver/tests/test_push_notifications.py

```diff
@@ -1483,6 +1483,56 @@ def test_user_message_soft_deactivated(self) -> None:
             mock_send_android.assert_called_with(user_identity, android_devices, {"gcm": True}, {})
             mock_push_notifications.assert_called_once()
 
+    def test_user_push_soft_reactivate_soft_deactivated_user(self) -> None:
+        sender = self.example_user("iago")
+        logger_string = "zulip.soft_deactivation"
+        mention_message = f"@**{self.user_profile.full_name}**"
+
+        with self.assertLogs(logger_string, level="INFO") as info_logs:
+            do_soft_deactivate_users([self.user_profile])
+        self.assertEqual(
+            info_logs.output,
+            [
+                f"INFO:{logger_string}:Soft deactivated user {self.user_profile.id}",
+                f"INFO:{logger_string}:Soft-deactivated batch of 1 users; 0 remain to process",
+            ],
+        )
+        self.assertTrue(self.user_profile.long_term_idle)
+
+        event = {"type": "soft_reactivate", "user_profile_id": self.user_profile.id}
+
+        message_id = self.send_stream_message(sender, "Denmark", mention_message)
+        missed_message = {
+            "message_id": message_id,
+            "trigger": "mentioned",
+        }
+        with mock_queue_publish(
+            "zerver.lib.push_notifications.queue_json_publish"
+        ) as patched, mock.patch(
+            "zerver.lib.push_notifications.push_notifications_enabled", return_value=True
+        ) as mock_push_notifications:
+            handle_push_notification(self.user_profile.id, missed_message)
+            patched.assert_called_once()
+            self.assertEqual(patched.call_args_list[0][0][0], "deferred_work")
+            self.assertEqual(patched.call_args_list[0][0][1], event)
+            mock_push_notifications.assert_called_once()
```

Comment:
I think I'd prefer a more end-to-end test here -- as it is, we never run the code to actually soft-reactivate the user. The test system will automatically call queue processors in `queue_json_publish` if we don't mock it, and I think that default behavior is better than this bit of verification of what goes into the queue.

https://zulip.readthedocs.io/en/latest/testing/philosophy.html has some higher level discussion of why we prefer end-to-end tests.

Reply:
I haven't added the E2E test yet, as I'm not sure if I understand how the test system works with the queue processors. Looking at the examples in `zerver/tests/test_queue_workers.py`, it appears to me a fake client needs to be set up, and events need to be explicitly enqueued for that client (so patching `queue_json_publish` might be required to swap in the fake client?)

Reply:
I think we talked about this, but `test_queue_workers` is code specifically for testing the queue worker framework, where that sort of mocking is useful. The `queue_json_publish` function itself handles calling the corresponding queue worker directly with the event (basically skipping the queueing framework), which is what we want for the vast majority of tests.